### ISSUE TITLE
Make template dependency errors more clear

### ISSUE THREAD

**Issue by Austinauth** [Azure/azure-dev#1449]
# Description

When attempting to use azd without the correct dependencies installed users may encounter unhelpful and difficult to parse error messages. This makes it hard for users troubleshoot in-context, often resulting in the inability to complete their task.

## Goals
- Design an updated pattern for dependency errors. Templates need a common (and human legible) error state when they don't have a dependency installed
- Update the docs to make it more clear templates may have their own dependancies

### Tasks
- [x] Capture real life examples of these dependency errors (@vhvb1989)
- [x] Identify the different types of dependency errors that need their own pattern 
- [x] Create sketches for each type of dependency pattern
- [x] Review the patterns with product & design team and finalize

### Running azd without Git installed
![image](https://user-images.githubusercontent.com/113634166/222276368-2b32ec28-afcc-4c04-9681-d20c10c3d42a.png)

![image](https://user-images.githubusercontent.com/113634166/222276759-fe8c4d5a-2ad2-419b-8651-119ae483c4ab.png)

Git dependency should show before providing template options.

### Running pipeline config without GitHub installed

![image](https://user-images.githubusercontent.com/113634166/222277135-2f476d7a-fd3f-4cfc-9e76-23064efbf7a3.png)

For pipeline config we also want to check for git before running. Don't let user proceed. 

### Terraform deployment without terraform installed
[Need screenshot]

### Running azd up on a .NET template w/out the dependency installed
![image](https://user-images.githubusercontent.com/113634166/222277693-b59a12c8-f0c2-4e4a-926f-764c9e98cac5.png)

![image](https://user-images.githubusercontent.com/113634166/221976031-d027e206-148f-4c65-9b5e-65580a6f3077.png)

- .NET only a requirement for templates using .NET
- .NET error only during provision or azd up
- Error from azd up and provison should be the same
- .NET depency should look like the GitHub one

**Comment by Austinauth** [comment]
![image](https://user-images.githubusercontent.com/6994218/214418664-3f0420ea-3052-438a-a94f-9efc47f9000b.png)

**Comment by savannahostrowski** [comment]
Related to #243

**Comment by rajeshkamal5050** [comment]
Adding @vhvb1989 to get plugged in as part of the design.

**Comment by rajeshkamal5050** [comment]
@Austinauth @vhvb1989 @savannahostrowski 

Since, we are talking about template dependencies. Can we also factor in design for root and command level dependency errors?

**Comment by Austinauth** [comment]
@rajeshkamal5050 are you able to provide examples for me to reference?

**Comment by rajeshkamal5050** [comment]
> @rajeshkamal5050 are you able to provide examples for me to reference?

Don't have the exact steps but. Trying todo the following should surface things up,
- Root level. Try running azd without git installed @vhvb1989 
- Command level. Try running
  - GitHub pipeline config without GitHub CLI installed https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/configure-devops-pipeline?tabs=GitHub#authorize-github-to-deploy-to-azure @vhvb1989 
  - Terraform deployment without Terraform installed https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/use-terraform-for-azd @wbreza

**Comment by savannahostrowski** [comment]
Here's an example:


![Image](https://user-images.githubusercontent.com/8949415/220441718-419a7394-3f88-4664-82cf-08125fc1ac6f.png)

I ran provision and I hadn't installed dotnet. I am missing the runtime and it's buried in this error message.

**Comment by savannahostrowski** [comment]
@wbreza/@vhvb1989/@rajeshkamal5050 Why does a template dependency error for dotnet show up on provision? I'd expect this to show up on deploy as part of the build step before the deploy. Any insight here? It's a little confusing.

**Comment by vhvb1989** [comment]
@savannahostrowski , `dotnet` became a requirement for `provisioning` .NET after this: https://github.com/Azure/azure-dev/issues/405

After `provisioning`, we set all bicep outputs as secrets in `dotnet`.  So we are failing early here while we initialize the dotnet-secrets before provisioning

**Comment by savannahostrowski** [comment]
aha! That's interesting. I wonder if it'd be worth trying to convey some of that (specifically for dotnet) in the error message. 

At first glance, I was confused.

**Comment by vhvb1989** [comment]
yes, @savannahostrowski , specially, I think we should update the Readme in NET templates and mention that dotnet is required for provisioning as well (not only deploy).  As I don't see this was added in the PR that introduced the project initialization logic

**Comment by Petermarcu** [comment]
What are the types of template dependencies we will check for? Is it primarily language/runtime support? Do we have a list of the ones we would check for centrally or a design for how we enable templates to express additional dependencies?

**Comment by Austinauth** [comment]
**Update 03/02/2023**

- Met with Victor yesterday to review examples
- [Initial comment](https://github.com/Azure/azure-dev/issues/1449#issue-1555234203) updated with screenshots and additional info
- Should have a design concept ready to share by EoD that covers that .NET, GitHub and Terraform scenarios. Should work for most other dependencies as well.

**Comment by Austinauth** [comment]
@rajeshkamal5050, @vhvb1989, @savannahostrowski, @wbreza 

Here are two approaches we could take for dependency errors. **If there is no preference or additional feedback let's assume we roll with option 1.**

Both options showcase how the message could scale if multiple dependencies are detected. In most scenarios it's likely that only one dependency will be required. While slightly harder to parse, I'm leaning option 1 because it is most consistent with our existing patterns.


## Option 1
![image](https://user-images.githubusercontent.com/6994218/222549470-16a5caa7-2516-4c98-a300-a6b7e75d92cb.png)

## Option 2
![image](https://user-images.githubusercontent.com/6994218/222549501-d6e02426-278c-4058-8747-90379af5e28f.png)

**Comment by savannahostrowski** [comment]
I think we want to distinguish between core tool missing dependencies vs. template dependencies. Can we categorize those?

**Comment by Petermarcu** [comment]
Yeah, I was just thinking, if we don't have git, we don't even know the template needs .Net

**Comment by Austinauth** [comment]
@savannahostrowski, @Petermarcu good catch. 

So if I'm understanding correctly:
- There are two primary types of dependancies: Tool (required for azd) and Template (required for the template)
- It could be valuable for the user to know if a dependency is one or the other for troubleshooting purposes.
- Git is a core tool dependency, if you don't have it we can't determine which template dependancies are missing. Therefore its unlikely we ever have a situation where we surface BOTH a tool and template dependency error.
- However, it is possible for the user to be missing multiple template dependencies. Therefore it's possible we would have to surface something like: Missing dependency - dotNet, Missing dependency -Node.js


Here's a how that could look:

## Missing core tool dependency

![image](https://user-images.githubusercontent.com/6994218/222777338-65a0503b-04ef-43bb-abdd-1125db21dba7.png)

## Missing template dependency
![image](https://user-images.githubusercontent.com/6994218/222777378-2351aca1-667e-43d1-afbc-4768bd6abe38.png)

**Comment by savannahostrowski** [comment]
Yep - that sounds about right.

That said, a user could be missing both core azd deps + template deps. In this case, they might experience:
- Run `azd init`
- Tool yells about missing git/see error
- Fix missing tool dep and install git
- Run `azd init` again
- Tool yells about missing template deps (e.g. node.js/dotnet)/see error
- Fix missing template deps
- Run `azd init`

**Comment by Austinauth** [comment]
@savannahostrowski 

Ah yep, gotcha. In that scenario there's no way to know ahead of time that template dependencies will be missing, right? So there's not really a way to optimize for this technically? We could show both types of errors at once _if_ that was possible.

**Comment by savannahostrowski** [comment]
That's my understanding.

I think that the flow I outlined above is acceptable. I think there's a fine line here where if there's too much information (e.g. lots of missing template deps), it could be hard to parse. I don't disagree that it feels nicer with this example to have all missing deps surfaced at once though.

**Comment by Petermarcu** [comment]
Do we have a feature for how templates declare their dependencies, how to check for them, and what the error message should be of missing? I know we have mainline template dependencies like language but many templates will depend on tools we don't have knowledge of.

**Comment by Austinauth** [comment]
> Do we have a feature for how templates declare their dependencies, how to check for them, and what the error message should be of missing? I know we have mainline template dependencies like language but many templates will depend on tools we don't have knowledge of. Also, not having the right version of a dependency will cause an error later. Do we need to know what version we depend on and how to check for that?

I can't answer your first question. However for the second one, are you suggesting we surface what version of the dependency is required? I think thats a good idea if we have that info and can tweak the design to accommodate. I'll let one of the devs chime in on whether its possible.

**Comment by Petermarcu** [comment]
It would be good to have an error for when the version is the issue to use. I think the answer to a lot of my questions is that we don't have all these features but I think we will want them eventually and having the UX accommodate that in the design would be great.

**Comment by Austinauth** [comment]
Ah yes, I understand now. Really appreciate the feedback. I can absolutely include in a variation of the error where we highlight the version # as the issue. I ran into this issue myself not long ago.

**Comment by rajeshkamal5050** [comment]
> Do we have a feature for how templates declare their dependencies, how to check for them, and what the error message should be of missing?

No, we don't. Feeling this falls under the extensibility bucket? @wbreza @ellismg 

> I know we have mainline template dependencies like language but many templates will depend on tools we don't have knowledge of. Also, not having the right version of a dependency will cause an error later. Do we need to know what version we depend on and how to check for that?

We should be careful with making strict version checks. Developers will have different versions of dependencies. Should we be checking for version above our minimum required?

We should also be flexible in making those versions configurable. So, that if any issues caused by our dependencies. CSS should be able to help/guide them to configure and move forward accordingly?

**Comment by Petermarcu** [comment]
Only min version as dictated by the template author IMO. Since we don't have these features, the best we can do right now is consider how the UX should look when we add them so we don't have to come back and revisit. But if it's too unknown, we can just do the simple thing we know about today.

**Comment by Austinauth** [comment]
@vhvb1989 I posted this in our UX group chat, but sounds like we have the 👍🏼 for the following. Please let me know if you have any questions or additional feedback.

## Missing core tool dependency

![image](https://user-images.githubusercontent.com/6994218/223220098-4c242ce4-ecb2-4de8-aa10-2819c929c930.png)

**If dependency is installed, but outdated:**
![image](https://user-images.githubusercontent.com/6994218/223219996-9570479b-3f1d-4a17-bbb8-2371441fa712.png)

The user's current version is added in parenthesis next to the version required.

## Missing template dependency
It was determined in our group chat that template dependency errors should not happen during the `init` step, rather `provision` or `deploy`.

**It is still unclear to me if we will be able to detect the dependency immediately or if certain provisioning steps might happen before detection.**

If its the later the error would look something like this:
![image](https://user-images.githubusercontent.com/6994218/223220830-5149a929-d3b3-40dc-8e9e-81cd964746a1.png)

This example shows one out of date dependency and one that is entirely missing.


cc: @vhvb1989, @wbreza, @savannahostrowski, @rajeshkamal5050

**Comment by rajeshkamal5050** [comment]
@vhvb1989 please take a look at the above mockups. Will keep it for April fit-n-finish iteration.

**Comment by rajeshkamal5050** [comment]
@vhvb1989 anything pending here which needs to be fixed for GA?

**Comment by rajeshkamal5050** [comment]
Moving this out of GA. Since most of the UX fit-n-finish items/improvements around template errors, dependency errors, services errors at this point is looking good from GA POV.

@savannahostrowski @ellismg

**Comment by rajeshkamal5050** [comment]
@vhvb1989 can you capture the remaining items on this issue. Cosmetic UX improvements which can be addressed for GA. Let us not do any new functional changes.

**Comment by vhvb1989** [comment]
No pending items

**Comment by savannahostrowski** [comment]
@vhvb1989 These mocks were implemented?
